File: go/pkg/service/service.go
package service

import (
	"fmt"

	"kclone-go/pkg/models"
	"kclone-go/pkg/state"
)

// CreateService creates a new service
func CreateService(cs *state.ClusterState, name string, selector map[string]string, port, targetPort int) (*models.Service, error) {
	vip := cs.AllocateVirtualIP()
	svc := &models.Service{
		Name:       name,
		Selector:   selector,
		Port:       port,
		TargetPort: targetPort,
		VirtualIP:  vip,
		Endpoints:  []string{},
		RRIndex:    0,
	}
	err := cs.AddService(svc)
	if err != nil {
		return nil, err
	}
	cs.RefreshServiceEndpoints()
	return svc, nil
}

// RouteRequest performs round-robin routing to a pod
func RouteRequest(cs *state.ClusterState, serviceName string) (string, error) {
	svc, exists := cs.GetService(serviceName)
	if !exists {
		return "", fmt.Errorf("service %s not found", serviceName)
	}

	cs.RefreshServiceEndpoints()

	if len(svc.Endpoints) == 0 {
		return "", fmt.Errorf("service %s has no ready pods", serviceName)
	}

	podUID := svc.Endpoints[svc.RRIndex%len(svc.Endpoints)]
	svc.RRIndex = (svc.RRIndex + 1) % len(svc.Endpoints)

	return podUID, nil
}

File: python/src/kclone/service.py
from __future__ import annotations

from .models import Service
from .state import ClusterState


def create_service(state: ClusterState, name: str, selector: dict, port: int, target_port: int) -> Service:
    vip = state.allocate_virtual_ip()
    svc = Service(name=name, selector=selector, port=port, target_port=target_port, virtual_ip=vip)
    state.add_service(svc)
    state.refresh_service_endpoints()
    return svc


def route_request(state: ClusterState, service_name: str) -> str:
    # Refresh endpoints before routing
    state.refresh_service_endpoints()
    svc = state.services.get(service_name)
    if not svc:
        raise KeyError(f"Service {service_name} not found")
    if not svc.endpoints:
        raise RuntimeError(f"Service {service_name} has no ready pods")
    pod_uid = svc.endpoints[svc.rr_index % len(svc.endpoints)]
    svc.rr_index = (svc.rr_index + 1) % len(svc.endpoints)
    return pod_uid
